libkbfs/tlf_journal.go

```diff
@@ -375,6 +376,134 @@ func (j *tlfJournal) resumeBackgroundWork() {
 	}
 }
 
+func (j *tlfJournal) getBlockRangeToFlush(ctx context.Context) (
+	doFlush bool, first journalOrdinal, last journalOrdinal, err error) {
+	j.journalLock.RLock()
+	defer j.journalLock.RUnlock()
+
+	length, err := j.blockJournal.length()
+	if err != nil {
+		return false, 0, 0, err
+	}
+	if length == 0 {
+		return false, 0, 0, nil
+	}
+
+	first, last, err = j.blockJournal.ordinalRange()
+	if err != nil {
+		return false, 0, 0, err
+	}
+	return true, first, last, nil
+}
+
+func (j *tlfJournal) getBlockEntryToFlush(ctx context.Context,
+	ordinal journalOrdinal) (*blockJournalEntry, []byte,
+	BlockCryptKeyServerHalf, error) {
+	j.journalLock.RLock()
+	defer j.journalLock.RUnlock()
+	return j.blockJournal.getEntryToFlush(ctx, ordinal)
+}
+
+func (j *tlfJournal) removeFlushedBlockEntry(ctx context.Context,
+	ordinal journalOrdinal, entry blockJournalEntry) (int64, error) {
+	j.journalLock.Lock()
+	defer j.journalLock.Unlock()
+	return j.blockJournal.removeFlushedEntry(ctx, ordinal, entry)
+}
+
+func (j *tlfJournal) flushBlocks(ctx context.Context,
+	first journalOrdinal, last journalOrdinal) (err error) {
+	j.flushLock.Lock()
+	defer j.flushLock.Unlock()
+
+	j.log.CDebugf(ctx, "Flushing block entries %d to %d", first, last)
+	defer func() {
+		j.log.CDebugf(ctx, "Finished flushing block entries: %v", err)
+	}()
+
+	putBps := newBlockPutState(int(last-first) + 1)
+	addRefBps := newBlockPutState(int(last-first) + 1)
+	var allEntries []*blockJournalEntry
+	var otherEntries []*blockJournalEntry
+
+	for ordinal := first; ordinal <= last; ordinal++ {
```

Comment:
That sounds ok, except that it could severely cut down on parallelism to force all of the entries returned from `getNextEntriesToFlush` to all be parallelizable or not.  There could be puts/addRefs waiting both before and after an archiveRef that could benefit from parallelization.  What if `getNextEntriesToFlush` returned a struct that contains a put `blockPutState`, an addRef `blockPutState`, and then a list of other entries that need to be done one-by-one?  That would be closer to what I have here, just moving most of the logic into `blockJournal`.